### PR TITLE
cache: refactor memcached client code

### PR DESCRIPTION
**Pull request by ortuman**
**What this PR does**:

This PR is a follow-up of https://github.com/grafana/dskit/pull/245, based on suggested feedback. 

It aims to extract generic memcached behavior, so that it can be reused on alternate remote cache implementations such as Redis. 

**Which issue(s) this PR fixes**:

Fixes N/A

**Checklist**
- [X] Tests updated
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: cache/memcached_client.go [hunk 1]
```diff
@@ -587,47 +512,23 @@ func (c *memcachedClient) trackError(op string, err error) {
...
-			op()
-		case <-c.stop:
-			return
-		}
+		c.metrics.failures.WithLabelValues(op, reasonOther).Inc()
 	}
 }
 
 func (c *memcachedClient) resolveAddrsLoop() {
-	defer c.workers.Done()
```
Comment (pracucci) [inline cache/memcached_client.go:629]: We lost this logic. We're not waiting for `resolveAddrsLoop()` go routine to end when we call `memcachedClient.Stop()`. However, I think it's not a problem.

File: cache/async_queue.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+		return errAsyncQueueFull
+	}
+}
+
+func (q *asyncQueue) stop() {
+	close(q.stopCh)
+	q.workers.Wait()
+}
+
+func (q *asyncQueue) length() int {
```
Comment (pracucci) [inline cache/async_queue.go:46]: What is this used for?

File: cache/async_queue_test.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+	var i atomic.Int32
+	_ = q.run(func() { i.Add(1) })
+	_ = q.run(func() { i.Add(-1) })
+	_ = q.run(func() { i.Add(1) })
+	_ = q.run(func() { i.Add(-1) })
+	_ = q.run(func() { i.Add(1) })
+
+	// Wait for all operations to finish.
+	time.Sleep(100 * time.Millisecond)
```
Comment (pracucci) [inline cache/async_queue_test.go:24]: Better to call `q.stop()` here. No need to `time.Sleep()` and we also test `stop()` waits until all workers have done.

File: cache/async_queue_test.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+	doneCh := make(chan struct{})
+	defer func() {
+		close(doneCh)
+	}()
+
+	// Keep worker busy.
+	_ = q.run(func() {
+		<-doneCh
+	})
+	time.Sleep(100 * time.Millisecond)
```
Comment (pracucci) [inline cache/async_queue_test.go:44]: Why is this required? Shouldn't be.

**Review by pracucci — APPROVED** [review]
LGTM (modulo a couple of nits). Thanks!

File: cache/remote_cache.go [hunk 5]
```diff
@@ -12,51 +12,66 @@ import (
 
 // MemcachedCache is a memcached-based cache.
 type MemcachedCache struct {
```
Comment (56quarters) [inline cache/remote_cache.go:14]: I'm unclear on why `MemcachedCache` exists now if it's exposing the same functionality as `remoteCache` after this refactor.

File: cache/async_queue.go [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+	// Start a number of goroutines - processing async operations - equal
+	// to the max concurrency we have.
+	q.workers.Add(maxConcurrency)
+	for i := 0; i < maxConcurrency; i++ {
+		go q.asyncQueueProcessLoop()
+	}
+	return q
+}
+
+func (q *asyncQueue) run(op func()) error {
```
Comment (56quarters) [inline cache/async_queue.go:32]: nit: I would call this `submit()` since it doesn't run the provided closure immediately.

File: cache/async_queue_test.go (hunk 3 above)
Comment (56quarters) [reply cache/async_queue_test.go:24]: I think your suggestion exposes a potential problem with the current implementation: not everything that has been queued will be executed if `.stop()` is called immediately due to the `select` statement in the worker loop.

File: cache/async_queue.go (hunk 6 above)
Comment (pracucci) [reply cache/async_queue.go:32]: Good suggestion!

File: cache/remote_cache.go (hunk 5 above)
Comment (pracucci) [reply cache/remote_cache.go:14]: I agree we should simplify the design of the cache client, but I would suggest to postpone it to a follow up PR.

File: cache/async_queue_test.go (hunk 3 above)
Comment (pracucci) [reply cache/async_queue_test.go:24]: Right, it's a valid concern.

File: cache/async_queue_test.go (hunk 4 above)
Comment (ortuman) [reply cache/async_queue_test.go:44]: Here we try to simulate a long running task that prevents our only worker from pulling any enqueued operation. This way we can fill the queue until getting the expected error. 

The sleep before the loop is done in order to ensure that the worker is already running before enqueuing any extra ops. Otherwise we may end up getting a _queue full error_ in the case it haven't triggered yet.

File: cache/remote_cache.go (hunk 5 above)
Comment (ortuman) [reply cache/remote_cache.go:14]: See: https://github.com/grafana/dskit/pull/245#pullrequestreview-1220550728
